File: python/bob/bob.py
def hey(phrase):

    last_non_space_character = None
    for i in range(len(phrase)):
        if not phrase[i].isspace():
            last_non_space_character = phrase[i]

    if last_non_space_character == '?':
        if phrase.isupper():
            return 'Calm down, I know what I\'m doing!'
        else:
            return 'Sure.'

    elif phrase.isupper():
        return 'Whoa, chill out!'

    elif not phrase or phrase.isspace():
        return 'Fine. Be that way!'
    else:
        return 'Whatever.'

File: clojure/bob/src/bob.clj
(ns bob
  (:require [clojure.string :as str]))

(defn upper-case? [s]
  (and (= s (str/upper-case s)) (not (= s (str/lower-case s)))))

(defn response-for [s]
  (let [phrase (str/trim s)]
    (cond
      (str/blank? phrase) "Fine. Be that way!"
      (and (str/ends-with? phrase "?") (upper-case? phrase)) "Calm down, I know what I'm doing!"
      (str/ends-with? phrase "?") "Sure."
      (upper-case? phrase) "Whoa, chill out!"
      :else "Whatever.")))
